userspace/libsinsp/test/filter_compiler.ut.cpp

```diff
@@ -0,0 +1,110 @@
+{
+public:
+	inline gen_event_filter *new_filter() override
+	{
+		return new sinsp_filter(NULL);
+	}
+
+	inline gen_event_filter_check *new_filtercheck(const char *fldname) override
+	{
+		return new mock_compiler_filter_check();
+	}
+
+	inline list<gen_event_filter_factory::filter_fieldclass_info> get_fields() override
+	{
+		return m_list;
+	}
+
+	list<gen_event_filter_factory::filter_fieldclass_info> m_list;
+};
+
+// Compile a filter, pass a mock event to it, and
+// check that the result of the boolean evaluation is
+// the expected one
+void test_filter_run(bool result, string filter_str)
+{
+	std::shared_ptr<gen_event_filter_factory> factory;
+	factory.reset(new mock_compiler_filter_factory());
+	sinsp_filter_compiler compiler(factory, filter_str);
+	try
+	{
+		auto filter = compiler.compile();
+		if (filter->run(NULL) != result)
+		{
+			FAIL() << filter_str << " -> unexpected '" << (result ? "false" : "true") << "' result";
+		}
+		delete filter;
+	}
+	catch(const sinsp_exception& e)
+	{
+		FAIL() << filter_str << " -> " << e.what();
+	}
+}
+
+// In each of these test cases, we compile filter expression
+// of which we can control the truth state of each filtercheck,
+// so that we can deterministically check the result of running
+// a mock event in the compiled filters. The purpose is verifying
+// that the compiler constructs valid boolean expressions.
+TEST(sinsp_filter_compiler, smoke_filter_run)
+{
+	test_filter_run(true,  "c.true=1");
+	test_filter_run(false, "c.false=1");
+	test_filter_run(false, "not c.true=1");
+	test_filter_run(true,  "not not c.true=1");
+	test_filter_run(true,  "not (not c.true=1)");
+	test_filter_run(false, "not not not c.true=1");
+	test_filter_run(false, "not (not (not c.true=1))");
+	test_filter_run(true,  "not not not not c.true=1");
+	test_filter_run(true,  "c.true=1 and c.true=1");
+	test_filter_run(false, "c.true=1 and c.false=1");
+	test_filter_run(false, "c.false=1 and c.true=1");
+	test_filter_run(false, "c.false=1 and c.false=1");
+	test_filter_run(false, "c.true=1 and not c.true=1");
+	test_filter_run(false, "not c.true=1 and c.true=1");
+	test_filter_run(true,  "c.true=1 or c.true=1");
+	test_filter_run(true,  "c.true=1 or c.false=1");
+	test_filter_run(true,  "c.false=1 or c.true=1");
+	test_filter_run(false, "c.false=1 or c.false=1");
+	test_filter_run(true,  "c.false=1 or not c.false=1");
+	test_filter_run(true,  "not c.false=1 or c.false=1");
+	test_filter_run(true,  "c.true=1 or c.true=1 and c.false=1");
+	test_filter_run(false, "(c.true=1 or c.true=1) and c.false=1");
+	test_filter_run(true,  "not (not (c.true=1 or c.true=1) and c.false=1)");
+	test_filter_run(false, "not (c.false=1 or c.false=1 or c.true=1)");
+	test_filter_run(true,  "not (c.false=1 or c.false=1 and not c.true=1)");
+	test_filter_run(false, "not (c.false=1 or not c.false=1 and c.true=1)");
+	test_filter_run(false, "not ((c.false=1 or not (c.false=1 and not c.true=1)) and c.true=1)");
```

Comment:
This looks like a great set of tests.  You *might* already be covered, but I wonder if these tests cover the ambiguous cases where logical operators are joined without parentheses. I bet there's a lot of that in the default falco rules files and we definitely want the end behavior to be the same, even if the old behavior was ambiguous.

Reply:
I'd also add some tests involving quoted strings and generally more complicated right hand sides than just "1".  I do remember several pitfalls where the right hand side of a comparison was going down the wrong path of the grammar--stuff like string with a leading digit being treated as numbers, pitfalls involving quoted strings, etc.

I know some heavy users of exceptions are being very aggressive putting fairly long complicated quoted strings for values like `proc.cmdline`, and this will certainly find those corner cases/bugs in the grammar.

Reply:
Hey Mark, thank you for your comments!
1. Note that both that the case with mixed `or` and `and` are fairly covered in both `filter_parser.ut.cpp` and `filter_compiler.ut.cpp`. This grammar support operator priority, just like the one in Falco, so `and` can always be mixed with `or` and `not` without ambiguity.
2. Note that performing real filterchecks is not the point of the unit tests in `filter_compiler.ut.cpp`. In that, filterchecks are mocked to return true or false deterministically (`c.false` is always false, `c.true` is always true), and the `<check>=1` is just a mocked check syntax that allows the filter to be parsable by the grammar. The goal of this is to make sure that the compiler generates a valid boolean evaluation tree. We do this by writing simple and predictable boolean formulas filters, and then check if they are evaluated with the expected true/false result once compiled.
3. I agree we may want to test escaping and single filtercheck implementation. Currently, those tests are totally missing in `master` too. In this case, I think it would be more valuable to try to make Falco run with this new parser and see if its integration tests pass. Will work on that.

Reply:
Thanks! Those tests really expand the code coverage.